src/lib/getCacher.ts

```diff
@@ -0,0 +1,98 @@
+import fs from 'fs'
+import os from 'os'
+import path from 'path'
+import { CacheData, Cacher } from '../types/Cacher'
+import { Options } from '../types/Options'
+import { print } from './logging'
+
+/**
+ * Check if cache is expired if timestamp is set
+ *
+ * @param cacheData
+ * @param cacheExpiration
+ * @returns
+ */
+function checkCacheExpiration(cacheData: CacheData, cacheExpiration = 10) {
+  if (typeof cacheData.timestamp !== 'number') {
+    return false
+  }
+
+  const unixMinuteMS = 60 * 1000
+  const expirationLimit = cacheData.timestamp + cacheExpiration * unixMinuteMS
+  return expirationLimit < Date.now()
+}
+
+export const defaultCacheFilename = '.ncu-cache.json'
+export const defaultCacheFile = `~/${defaultCacheFilename}`
+export const resolvedDefaultCacheFile = path.join(os.homedir(), defaultCacheFilename)
+
+/**
+ * The cacher stores key (name + version) - value (new version) pairs
+ * for quick updates across `ncu` calls.
+ *
+ * @returns
+ */
+export default function getCacher(options: Omit<Options, 'cacher'>): Cacher | undefined {
+  if (!options.cache) {
+    return
+  }
+
+  const file = options.cacheFile
+  if (!file) {
+    return
+  }
+  const cacheFile = file === defaultCacheFile ? resolvedDefaultCacheFile : file
+  let cacheData: CacheData = {}
+  let cacheUsed: string[] = []
+
+  try {
+    cacheData = JSON.parse(fs.readFileSync(cacheFile, 'utf-8'))
+
+    const expired = checkCacheExpiration(cacheData, options.cacheExpiration)
+    if (expired) {
+      // reset cache
+      fs.promises.rm(cacheFile)
+      cacheData = {}
+    }
+  } catch (error) {
+    // ignore file read/parse/remove errors
+  }
+
+  if (typeof cacheData.timestamp !== 'number') {
+    cacheData.timestamp = Date.now()
+  }
+  if (!cacheData.packages) {
+    cacheData.packages = {}
+  }
+
+  return {
+    key: (name, version) => name + '@' + version,
+    get: key => {
+      if (!key || !cacheData.packages) return
+      const cached = cacheData.packages[key]
+      if (cached) {
+        cacheUsed = [...cacheUsed, key + `: ${cached}`]
+      }
+      return cached
+    },
+    set: (key, value) => {
+      if (!key || !cacheData.packages) return
+      cacheData.packages[key] = value
+    },
+    save: async () => {
+      await fs.promises.writeFile(cacheFile, JSON.stringify(cacheData))
+    },
+    log: () => {
+      if (cacheUsed.length === 0) return
+
+      const message = '\nUsing cached package versions'
+
+      if (options.loglevel === 'verbose') {
+        print(options, message + ':', 'verbose')
+        cacheUsed.forEach(cache => print(options, `  ${cache}`, 'verbose'))
```

Comment:
I have a small change I'd like to make to the verbose log. Let's output an object of cached versions. It would look like this:

```
Using 2 cached package versions:
{
  "react": "x.y.z"
  "typescript": "x.y.z"
}
```

Note that it should only show the cache hits, not the entire cache.

Reply:
By cache hit you mean a difference in package versions? The cacher caches everything, so a package which does not change `one@1.0.0` is also cached with the version `1.0.0` and future queries for the package `one` will also be a cache hit, even if the same version. This is what decreases the update time since it goes over redundant checks, and since the default cache expiration is only 10 minutes, there is a very small chance that the package was updated precisely in those 10 minutes.

Reply:
What I mean is, it should only show cached packages that have upgrades. If the user really needs to look at the entire cache (which I imagine would be rare), they can look at the cache file itself. Within the context of running ncu in verbose mode, we want to tell them exactly which upgrades came from the cache.

Reply:
Right. In the example from above in a React package we have:

<img width="529" alt="Screenshot 2022-09-04 at 10 10 56" src="https://user-images.githubusercontent.com/13864427/188301921-b21f2a00-d0f9-4c58-afc2-2b0b4920f2c2.png">

`react@^18.2.0` is also a cache hit, even if the 'new' version is the same version, `18.2.0`. You would like to display only `@types/react-dom` since it's a cache hit with a package upgrade, correct?

Reply:
Yes, exactly.